README.md
<img src="https://upload.wikimedia.org/wikipedia/commons/f/f2/Flag_of_Costa_Rica.svg" alt="imagen de la bandera de costa rica" width="200" height="200">

# Costa Rica IBAN
![Tests](https://github.com/jeanlescure/costa-rica-iban/workflows/tests/badge.svg)
<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
[![All Contributors](https://img.shields.io/badge/all_contributors-4-orange.svg?style=flat-square)](#contributors-)
<!-- ALL-CONTRIBUTORS-BADGE:END -->

Funciones de Typescript/Javascript útiles para extraer y validar información general de números de cuenta [IBAN de Costa Rica](https://www.bccr.fi.cr/sistema-de-pagos/informaci%C3%B3n-general/iban).

Esta herramienta fue creada en base al documento [Codificaciones Generales Del Sistema De Pagos](https://raw.githubusercontent.com/jeanlescure/costa-rica-iban/master/assets/codificaciones-generales-del-sistema-de-pagos.pdf)
del Banco Central de Costa Rica (BCCR). El Jupyter Notebook usado para la extracción de datos puede
ser visualizado en [este enlace](https://costaricaiban.org/assets/cotejando-codificaciones-generales-del-sistema-de-pagos.html).

## Herramienta en línea

Para validar números de cuenta IBAN y extraer información sobre las entidad bancaria emisora, puede hacerse uso de la siguiente página oficialmente mantenida por los colaboradores de este repositorio:

[https://costaricaiban.org](https://costaricaiban.org)

## Uso con Node.js

Instalación pnpm:

```
$ pnpm add costa-rica-iban
```

Instalación yarn:

```
$ yarn add costa-rica-iban
```

Instalación npm:

```
$ npm install costa-rica-iban
```

Uso de funciones:

```js
import {
  getBankNameFromIBAN,
  getBankObjectFromIBAN,
} from 'costa-rica-iban';


// Cuenta IBAN en entidad bancaria sin representante:
let bankAcct = 'CR06010200009123456789';

console.log(getBankNameFromIBAN(bankAcct));
// > BAC San José S.A.


// Cuenta IBAN en entidad bancaria con representante:
bankAcct = 'CR06083700009123456789';

console.log(getBankNameFromIBAN(bankAcct));
// > Credomatic

console.log(getBankNameFromIBAN(bankAcct), true);
// > BAC San José S.A.


console.log(getBankObjectFromIBAN(bankAcct));
// >
// {
//   "code": "837",
//   "entity": "Credomatic",
//   "category": "ENTIDAD LIQUIDADORA DE SERVICIOS Y EMISORA DE TARJETAS",
//   "participation": "representada",
//   "representative": "BAC San José S.A."
// }

```

Funciones disponibles:

- `getCountryPrefixFromIBAN(iban)`
- `verifyIBANCountryPrefix(iban)`
- `verifyIBANLength(iban)`
- `verifyIBANFormat(iban)`
- `getBankCodeFromIBAN(iban)`
- `getBankObjectFromIBAN(iban)`
- `getBankCategoryFromIBAN(iban)`
- `getBankNameFromIBAN(iban, returnRepresentative = false)`

Uso de clase:

```js
import CostaRicaIBAN from 'costa-rica-iban';

const cri = new CostaRicaIBAN('CR06010200009123456789');

console.log(cri.getBankObject());
// >
// {
//   "code": "102",
//   "entity": "BAC San José S.A.",
//   "participation": "asociada",
//   "representative": ""
// }
```

Funciones de clase disponibles:

- `.getCountryPrefix()`
- `.getBankCode()`
- `.getBankObject()`
- `.getBankCategory()`
- `.getBankName()`

Para más detalles ver las [pruebas](https://github.com/jeanlescure/costa-rica-iban/tree/master/src/index.test.js) en `./src/index.test.ts`.

## Uso en el navegador

```html
<script src="/dist/costa-rica-iban.min.js"></script>
<script>
  var cri = new CostaRicaIBAN('CR06010200009123456789');

  document.write(cri.getBankObject());
</script>
```

## Desarrollo

Clonar repositorio e instalar dependencias:

```
$ pnpm
```

Usamos [TDD](https://es.wikipedia.org/wiki/Desarrollo_guiado_por_pruebas), por lo tanto se espera que se agreguen las pruebas pertinentes en `./src/index.test.ts`.

Nuestro motor de pruebas es [Jest](https://jestjs.io/). Para ejecutar las pruebas simplemente hace falta correr el comando:

```
$ pnpm test
```

## Colaboración

Este repositorio tiene activados los issues y pull requests.

Para que un pull request sea tomado en cuenta deberá tener:

- un issue que describa la funcionalidad existente que esté rota o la funcionalidad nueva propuesta
- una descripción en el pull request que incluya mención del issue, un detalles de las soluciones/mejoras propuestas, y fuentes usadas para llegar a las soluciones/mejoras propuestas

## Colaboradores ✨

Gracias a los colaboradores de este proyecto ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://jeanlescure.cr"><img src="https://avatars2.githubusercontent.com/u/3330339?v=4" width="100px;" alt=""/><br /><sub><b>Jean Lescure</b></sub></a><br /><a href="https://github.com/jeanlescure/costa-rica-iban/commits?author=jeanlescure" title="Code">💻</a> <a href="#content-jeanlescure" title="Content">🖋</a> <a href="https://github.com/jeanlescure/costa-rica-iban/commits?author=jeanlescure" title="Documentation">📖</a> <a href="#design-jeanlescure" title="Design">🎨</a> <a href="#example-jeanlescure" title="Examples">💡</a> <a href="#maintenance-jeanlescure" title="Maintenance">🚧</a> <a href="https://github.com/jeanlescure/costa-rica-iban/commits?author=jeanlescure" title="Tests">⚠️</a> <a href="#userTesting-jeanlescure" title="User Testing">📓</a></td>
    <td align="center"><a href="https://dianalu.design"><img src="https://avatars2.githubusercontent.com/u/1036995?v=4" width="100px;" alt=""/><br /><sub><b>Diana Lescure</b></sub></a><br /><a href="#userTesting-DiLescure" title="User Testing">📓</a> <a href="#a11y-DiLescure" title="Accessibility">️️️️♿️</a></td>
    <td align="center"><a href="https://github.com/RoloGH"><img src="https://avatars2.githubusercontent.com/u/16967799?v=4" width="100px;" alt=""/><br /><sub><b>Rolando Gonzalez H.</b></sub></a><br /><a href="https://github.com/jeanlescure/costa-rica-iban/pulls?q=is%3Apr+reviewed-by%3ARoloGH" title="Reviewed Pull Requests">👀</a></td>
    <td align="center"><a href="https://kevinwolf.dev"><img src="https://avatars2.githubusercontent.com/u/3157426?v=4" width="100px;" alt=""/><br /><sub><b>Kevin Wolf</b></sub></a><br /><a href="#ideas-iamkevinwolf" title="Ideas, Planning, & Feedback">🤔</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->

Este proyecto sigue la especificación de [all-contributors](https://github.com/all-contributors/all-contributors). ¡Colaboraciones de cualquier tipo son bienvenidas!

## LICENCIA

Copyright 2020 Jean M. Lescure

Por la presente se concede permiso, libre de cargos, a cualquier persona que obtenga una copia de este software y de los archivos de documentación asociados (el "Software"), a utilizar el Software sin restricción, incluyendo sin limitación los derechos a usar, copiar, modificar, fusionar, publicar, distribuir, sublicenciar, y/o vender copias del Software, y a permitir a las personas a las que se les proporcione el Software a hacer lo mismo, sujeto a las siguientes condiciones:

El aviso de copyright anterior y este aviso de permiso se incluirán en todas las copias o partes sustanciales del Software.

EL SOFTWARE SE PROPORCIONA "COMO ESTÁ", SIN GARANTÍA DE NINGÚN TIPO, EXPRESA O IMPLÍCITA, INCLUYENDO PERO NO LIMITADO A GARANTÍAS DE COMERCIALIZACIÓN, IDONEIDAD PARA UN PROPÓSITO PARTICULAR E INCUMPLIMIENTO. EN NINGÚN CASO LOS AUTORES O PROPIETARIOS DE LOS DERECHOS DE AUTOR SERÁN RESPONSABLES DE NINGUNA RECLAMACIÓN, DAÑOS U OTRAS RESPONSABILIDADES, YA SEA EN UNA ACCIÓN DE CONTRATO, AGRAVIO O CUALQUIER OTRO MOTIVO, DERIVADAS DE, FUERA DE O EN CONEXIÓN CON EL SOFTWARE O SU USO U OTRO TIPO DE ACCIONES EN EL SOFTWARE.


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [2.0.0](https://github.com/jeanlescure/costa-rica-iban/compare/1.0.3...2.0.0)

- Bump hosted-git-info from 2.8.5 to 2.8.9 [`#18`](https://github.com/jeanlescure/costa-rica-iban/pull/18)
- Bump ssri from 6.0.1 to 6.0.2 [`#16`](https://github.com/jeanlescure/costa-rica-iban/pull/16)
- Bump lodash from 4.17.19 to 4.17.21 [`#17`](https://github.com/jeanlescure/costa-rica-iban/pull/17)
- Bump y18n from 4.0.0 to 4.0.1 [`#15`](https://github.com/jeanlescure/costa-rica-iban/pull/15)
- Bump elliptic from 6.5.3 to 6.5.4 [`#14`](https://github.com/jeanlescure/costa-rica-iban/pull/14)
- Bump ini from 1.3.5 to 1.3.8 [`#13`](https://github.com/jeanlescure/costa-rica-iban/pull/13)
- Bump elliptic from 6.5.2 to 6.5.3 [`#11`](https://github.com/jeanlescure/costa-rica-iban/pull/11)
- Bump lodash from 4.17.15 to 4.17.19 [`#10`](https://github.com/jeanlescure/costa-rica-iban/pull/10)
- Bump acorn from 6.4.0 to 6.4.1 [`#9`](https://github.com/jeanlescure/costa-rica-iban/pull/9)
- Bump yargs-parser from 13.1.1 to 13.1.2 [`#12`](https://github.com/jeanlescure/costa-rica-iban/pull/12)
- docs: add DiLescure as a contributor [`#8`](https://github.com/jeanlescure/costa-rica-iban/pull/8)
- docs: add iamkevinwolf as a contributor [`#7`](https://github.com/jeanlescure/costa-rica-iban/pull/7)
- docs: add RoloGH as a contributor [`#6`](https://github.com/jeanlescure/costa-rica-iban/pull/6)
- docs: add jeanlescure as a contributor [`#5`](https://github.com/jeanlescure/costa-rica-iban/pull/5)
- chore: extract data from pdf/csv using jupyter [`9940f5a`](https://github.com/jeanlescure/costa-rica-iban/commit/9940f5a3d2365fcebe5bff57d6218406662bcc3c)
- feat: implement getBankCategoryFromIBAN [`7b9791b`](https://github.com/jeanlescure/costa-rica-iban/commit/7b9791ba20508bc9aa053a6ffd999140adc494a7)
- chore: build dist for release [`acf8760`](https://github.com/jeanlescure/costa-rica-iban/commit/acf8760cc020f017bc8bce279007c78ff9b08769)

#### 1.0.3

> 14 February 2020

- Fix verifyIBANFormat false positive [`#3`](https://github.com/jeanlescure/costa-rica-iban/pull/3)
- initial commit [`d85b847`](https://github.com/jeanlescure/costa-rica-iban/commit/d85b8478ee535250bbeff7f96381049807b9b83f)
- add deployment process and github page [`fd735ad`](https://github.com/jeanlescure/costa-rica-iban/commit/fd735ad60e8c13e4e7ab6c0b39ce7e7b0a3d6449)
- implement get bank object test [`703ff1d`](https://github.com/jeanlescure/costa-rica-iban/commit/703ff1d40f0554a84f968f6535e68adc04e6d17b)


src/index.test.ts
import CostaRicaIBAN, {
  getCountryPrefixFromIBAN,
  verifyIBANCountryPrefix,
  verifyIBANLength,
  verifyIBANFormat,
  getBankCodeFromIBAN,
  getBankObjectFromIBAN,
  getBankNameFromIBAN,
  NOT_STRING_ERROR,
  INVALID_IBAN_ERROR,
  BankObject,
} from '.';

import bankCollection from './bank-collection.json';

const goodIBAN = 'CR06010200009123456789';
const goodIBANBank = bankCollection.find((b) => b.code === '102') as BankObject;

const badIBAN = 'DE69 5021 0900 0123 4567 13';

const deceptiveIBAN = 'CR06030500009123456789';

const ambiguousIBAN = 'CR06082400009123456789';
const ambiguousIBANBank = bankCollection.find((b) => b.code === '824') as BankObject;

describe('Costa Rica IBAN functions', () => {
  it('should be able to get the country prefix', () => {
    expect(
      // @ts-ignore
      getCountryPrefixFromIBAN
    ).toThrow(NOT_STRING_ERROR);

    expect(getCountryPrefixFromIBAN(goodIBAN)).toBe('CR');
  });

  it('should be able to verify the country prefix', () => {
    expect(
      // @ts-ignore
      verifyIBANCountryPrefix
    ).toThrow(NOT_STRING_ERROR);

    expect(verifyIBANCountryPrefix(badIBAN)).toBe(false);
    expect(verifyIBANCountryPrefix(goodIBAN)).toBe(true);
  });

  it('should be able to verify the length of the accounts', () => {
    expect(
      // @ts-ignore
      verifyIBANLength
    ).toThrow(NOT_STRING_ERROR);

    expect(verifyIBANLength(badIBAN)).toBe(false);
    expect(verifyIBANLength(goodIBAN)).toBe(true);
  });

  it('should be able to verify the correct format of the accounts', () => {
    expect(
      // @ts-ignore
      verifyIBANFormat
    ).toThrow(NOT_STRING_ERROR);

    expect(verifyIBANFormat(badIBAN)).toBe(false);
    expect(verifyIBANFormat(deceptiveIBAN)).toBe(false);
    expect(verifyIBANFormat(goodIBAN)).toBe(true);
  });

  it('should be able to get the bank code', () => {
    expect(
      // @ts-ignore
      () => getBankCodeFromIBAN(badIBAN)
    ).toThrow(INVALID_IBAN_ERROR);

    expect(getBankCodeFromIBAN(goodIBAN)).toBe(goodIBANBank.code);
  });

  it('should be able to get the bank object', () => {
    expect(
      // @ts-ignore
      () => getBankObjectFromIBAN(badIBAN)
    ).toThrow(INVALID_IBAN_ERROR);

    expect(getBankObjectFromIBAN(goodIBAN).entity).toBe(goodIBANBank.entity);
  });

  it('should be able to get the bank name', () => {
    expect(
      // @ts-ignore
      () => getBankNameFromIBAN(badIBAN)
    ).toThrow(INVALID_IBAN_ERROR);

    expect(getBankNameFromIBAN(goodIBAN)).toBe(goodIBANBank.entity);
    expect(getBankNameFromIBAN(ambiguousIBAN)).toBe(ambiguousIBANBank.entity);
    expect(getBankNameFromIBAN(ambiguousIBAN, true)).toBe(ambiguousIBANBank.representative);
  });
});

describe('Costa Rica IBAN class', () => {
  it('should be able to initialize with a valid Costa Rica IBAN', () => {
    expect(
      () => new CostaRicaIBAN(badIBAN)
    ).toThrow(INVALID_IBAN_ERROR);

    expect(
      // @ts-ignore
      () => new CostaRicaIBAN()
    ).toThrow(NOT_STRING_ERROR);

    const cri = new CostaRicaIBAN(goodIBAN);
    expect(cri.iban).toBe(goodIBAN);
  });

  it('should be able to get the country prefix', () => {
    const cri = new CostaRicaIBAN(goodIBAN);

    expect(cri.getCountryPrefix()).toBe('CR');
  });

  it('should be able to get the bank code', () => {
    const cri = new CostaRicaIBAN(goodIBAN);

    expect(cri.getBankCode()).toBe(goodIBANBank.code);
  });

  it('should be able to get the bank object', () => {
    const cri = new CostaRicaIBAN(goodIBAN);

    expect(cri.getBankObject().entity).toBe(goodIBANBank.entity);
  });

  it('should be able to get the bank name', () => {
    const cri = new CostaRicaIBAN(goodIBAN);

    expect(cri.getBankName()).toBe(goodIBANBank.entity);
  });

  it('should be able to get the bank name', () => {
    const cri = new CostaRicaIBAN(goodIBAN);

    expect(cri.getBankCategory()).toBe(goodIBANBank.category);
  });
});


test/criban_test.js
import CostaRicaIBAN, {
  getCountryPrefixFromIBAN,
  verifyIBANCountryPrefix,
  verifyIBANLength,
  verifyIBANFormat,
  getBankCodeFromIBAN,
  getBankObjectFromIBAN,
  getBankNameFromIBAN,
} from '../src';

const bankCollection = require('../src/bank-collection.json');

const goodIBAN = 'CR06010200009123456789';
const goodIBANBank = bankCollection.find((b) => b.code === '102');

const badIBAN = 'DE69 5021 0900 0123 4567 13';

const deceptiveIBAN = 'CR06030500009123456789';

const ambiguousIBAN = 'CR06082400009123456789';
const ambiguousIBANBank = bankCollection.find((b) => b.code === '824');

describe('Costa Rica IBAN functions', () => {
  it('should be able to get the country prefix', () => {
    try {
      getCountryPrefixFromIBAN();
      expect(true).toBe(false);
    } catch (e) {
      expect(e.message).toBe('Type Error: expected string');
      expect(getCountryPrefixFromIBAN(goodIBAN)).toBe('CR');
    }
  });

  it('should be able to verify the country prefix', () => {
    try {
      verifyIBANCountryPrefix();
      expect(true).toBe(false);
    } catch (e) {
      expect(e.message).toBe('Type Error: expected string');
      expect(verifyIBANCountryPrefix(badIBAN)).toBe(false);
      expect(verifyIBANCountryPrefix(goodIBAN)).toBe(true);
    }
  });

  it('should be able to verify the length of the accounts', () => {
    try {
      verifyIBANLength();
      expect(true).toBe(false);
    } catch (e) {
      expect(e.message).toBe('Type Error: expected string');
      expect(verifyIBANLength(badIBAN)).toBe(false);
      expect(verifyIBANLength(goodIBAN)).toBe(true);
    }
  });

  it('should be able to verify the correct format of the accounts', () => {
    try {
      verifyIBANFormat();
      expect(true).toBe(false);
    } catch (e) {
      expect(e.message).toBe('Type Error: expected string');
      expect(verifyIBANFormat(badIBAN)).toBe(false);
      expect(verifyIBANFormat(deceptiveIBAN)).toBe(false);
      expect(verifyIBANFormat(goodIBAN)).toBe(true);
    }
  });

  it('should be able to get the bank code', () => {
    try {
      getBankCodeFromIBAN(badIBAN);
      expect(true).toBe(false);
    } catch (e) {
      expect(e.message).toBe('Format Error: invalid Costa Rica IBAN format');
      expect(getBankCodeFromIBAN(goodIBAN)).toBe(goodIBANBank.code);
    }
  });

  it('should be able to get the bank object', () => {
    try {
      getBankObjectFromIBAN(badIBAN);
      expect(true).toBe(false);
    } catch (e) {
      expect(e.message).toBe('Format Error: invalid Costa Rica IBAN format');
      expect(getBankObjectFromIBAN(goodIBAN).entity).toBe(goodIBANBank.entity);
    }
  });

  it('should be able to get the bank name', () => {
    try {
      getBankNameFromIBAN(badIBAN);
      expect(true).toBe(false);
    } catch (e) {
      expect(e.message).toBe('Format Error: invalid Costa Rica IBAN format');
      expect(getBankNameFromIBAN(goodIBAN)).toBe(goodIBANBank.entity);
      expect(getBankNameFromIBAN(ambiguousIBAN)).toBe(ambiguousIBANBank.entity);
      expect(getBankNameFromIBAN(ambiguousIBAN, true)).toBe(ambiguousIBANBank.representative);
    }
  });
});

describe('Costa Rica IBAN class', () => {
  it('should be able to initialize with a valid Costa Rica IBAN', () => {
    let cri;

    try {
      cri = new CostaRicaIBAN(badIBAN);
      expect(true).toBe(false);
    } catch (e) {
      expect(e.message).toBe('Format Error: invalid Costa Rica IBAN format');
    }

    try {
      cri = new CostaRicaIBAN();
      expect(true).toBe(false);
    } catch (e) {
      expect(e.message).toBe('Type Error: expected string');

      cri = new CostaRicaIBAN(goodIBAN);
      expect(cri.iban).toBe(goodIBAN);
    }
  });

  it('should be able to get the country prefix', () => {
    const cri = new CostaRicaIBAN(goodIBAN);

    expect(cri.getCountryPrefix()).toBe('CR');
  });

  it('should be able to get the bank code', () => {
    const cri = new CostaRicaIBAN(goodIBAN);

    expect(cri.getBankCode()).toBe(goodIBANBank.code);
  });

  it('should be able to get the bank object', () => {
    const cri = new CostaRicaIBAN(goodIBAN);

    expect(cri.getBankObject().entity).toBe(goodIBANBank.entity);
  });

  it('should be able to get the bank name', () => {
    const cri = new CostaRicaIBAN(goodIBAN);

    expect(cri.getBankName()).toBe(goodIBANBank.entity);
  });
});


src/index.js
const { version } = require('../package.json');
const bankCollection = require('./bank-collection.json');

const bankCodes = bankCollection.map((b) => b.code);

const NOT_STRING_ERROR = new Error('Type Error: expected string');
const INVALID_IBAN_ERROR = new Error('Format Error: invalid Costa Rica IBAN format');

const errorIfNotString = (varToverify) => {
  if (typeof varToverify !== 'string') {
    throw NOT_STRING_ERROR;
  }
};

export const getCountryPrefixFromIBAN = (iban) => {
  errorIfNotString(iban);

  return iban.slice(0, 2);
};

export const verifyIBANCountryPrefix = (iban) => {
  errorIfNotString(iban);

  return getCountryPrefixFromIBAN(iban) === 'CR';
};

export const verifyIBANLength = (iban) => {
  errorIfNotString(iban);

  return iban.length === 22;
};

export const verifyIBANFormat = (iban) => {
  errorIfNotString(iban);

  return (/^CR\d\d0(1|[3-8])\d{16}$/).test(iban)
    && bankCodes.indexOf(iban.slice(5, 8)) > -1;
};

const errorOnInvalidIBAN = (iban) => {
  if (!verifyIBANFormat(iban)) {
    throw INVALID_IBAN_ERROR;
  }
};

export const getBankCodeFromIBAN = (iban) => {
  errorOnInvalidIBAN(iban);

  return iban.slice(5, 8);
};

export const getBankObjectFromIBAN = (iban) => {
  const bankCode = getBankCodeFromIBAN(iban);
  const bankIndex = bankCodes.indexOf(bankCode);

  return bankCollection[bankIndex];
};

export const getBankNameFromIBAN = (iban, returnRepresentative = false) => {
  const bankObject = getBankObjectFromIBAN(iban);

  return (returnRepresentative && bankObject.participation === 'representada') ? (
    bankObject.representative
  ) : (
    bankObject.entity
  );
};

class CostaRicaIBAN {
  iban = null;

  constructor(iban) {
    errorOnInvalidIBAN(iban);

    this.iban = iban;
  }

  getCountryPrefix() {
    return getCountryPrefixFromIBAN(this.iban);
  }

  getBankCode() {
    return getBankCodeFromIBAN(this.iban);
  }

  getBankObject() {
    return getBankObjectFromIBAN(this.iban);
  }

  getBankName() {
    return getBankNameFromIBAN(this.iban);
  }

  getVersion = () => version;
}

export default CostaRicaIBAN;


src/index.ts
// @ts-ignore
import { version } from '../package.json';
import bankCollection from './bank-collection.json';

const bankCodes: string[] = bankCollection.map((b) => b.code);

export interface BankObject {
  code: string;
  entity: string;
  category: string;
  participation: string;
  representative: string;
};

export const NOT_STRING_ERROR = new Error('Type Error: expected string');
export const INVALID_IBAN_ERROR = new Error('Format Error: invalid Costa Rica IBAN format');

// Since the browser version of the library cannot make use of types,
// we added this helper to make sure the input is a string.
const errorIfNotString = (varToverify: any) => {
  if (typeof varToverify !== 'string') {
    throw NOT_STRING_ERROR;
  }
};

export const getCountryPrefixFromIBAN = (iban: string) => {
  errorIfNotString(iban);

  return iban.slice(0, 2);
};

export const verifyIBANCountryPrefix = (iban: string) => {
  errorIfNotString(iban);

  return getCountryPrefixFromIBAN(iban) === 'CR';
};

export const verifyIBANLength = (iban: string) => {
  errorIfNotString(iban);

  return iban.length === 22;
};

export const verifyIBANFormat = (iban: string) => {
  errorIfNotString(iban);

  return (/^CR\d\d0(1|[3-8])\d{16}$/).test(iban)
    && bankCodes.indexOf(iban.slice(5, 8)) > -1;
};

const errorOnInvalidIBAN = (iban: string) => {
  if (!verifyIBANFormat(iban)) {
    throw INVALID_IBAN_ERROR;
  }
};

export const getBankCodeFromIBAN = (iban: string) => {
  errorOnInvalidIBAN(iban);

  return iban.slice(5, 8);
};

export const getBankObjectFromIBAN = (iban: string) => {
  const bankCode = getBankCodeFromIBAN(iban);
  const bankIndex = bankCodes.indexOf(bankCode);

  return bankCollection[bankIndex] as BankObject;
};

export const getBankNameFromIBAN = (iban: string, returnRepresentative = false) => {
  const bankObject = getBankObjectFromIBAN(iban);

  return (returnRepresentative && bankObject.participation === 'representada') ? (
    bankObject.representative
  ) : (
    bankObject.entity
  );
};

export const getBankCategoryFromIBAN = (iban: string, returnRepresentative = false) => {
  const bankObject = getBankObjectFromIBAN(iban);

  return bankObject.category;
};

class CostaRicaIBAN {
  iban: string;

  constructor(iban: string) {
    errorOnInvalidIBAN(iban);

    this.iban = iban;
  }

  getCountryPrefix() {
    return getCountryPrefixFromIBAN(this.iban);
  }

  getBankCode() {
    return getBankCodeFromIBAN(this.iban);
  }

  getBankObject() {
    return getBankObjectFromIBAN(this.iban);
  }

  getBankName() {
    return getBankNameFromIBAN(this.iban);
  }

  getBankCategory() {
    return getBankCategoryFromIBAN(this.iban);
  }

  getVersion = () => version;
}

export default CostaRicaIBAN;
